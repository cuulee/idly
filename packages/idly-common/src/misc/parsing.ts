export function replacer(_: string, value: any) {
  // Filtering out properties
  if (value instanceof Map) {
    return { '@@idly@type': 'Map', value: Array.from(value) };
  }
  if (value instanceof Set) {
    return { '@@idly@type': 'Set', value: Array.from(value) };
  }

  return value;
}

export function reviver(_: string, value: any) {
  if (
    value !== null &&
    typeof value === 'object' &&
    value['@@idly@type'] === 'Map'
  ) {
    return new Map(value.value);
  }

  if (
    value !== null &&
    typeof value === 'object' &&
    value['@@idly@type'] === 'Set'
  ) {
    return new Set(value.value);
  }
  return value;
}
