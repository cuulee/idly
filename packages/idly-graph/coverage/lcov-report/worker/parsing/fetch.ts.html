<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for worker/parsing/fetch.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">worker/parsing</a> fetch.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">77.78% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>28/36</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">90.91% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>10/11</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">77.78% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>7/9</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.33% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>25/30</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34</td><td class="line-coverage quiet"><span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">160x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-yes">240x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-yes">80x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { mercator } from 'idly-common/lib/geo/sphericalMercator';
import { handleErrors } from 'idly-common/lib/network/promise';
import { DOMParser } from 'xmldom';
<span class="cstat-no" title="statement not covered" ></span>
import { stubXML <span class="fstat-no" title="function not covered" >} from '.</span>./parsing<span class="cstat-no" title="statement not covered" >/fixtu<span class="cstat-no" title="statement not covered" >res';<span class="cstat-no" title="statement not covered" ></span></span></span>
&nbsp;
const fetchStub = () =&gt; {
  return Promise.resolve(stubXML);
};
&nbsp;
const debug = false;
&nbsp;
export async function fetchTile(x: number, y: number, zoom: number) {
  const xyz = [x, y, zoom].join(',');
  const bboxStr = <span class="fstat-no" title="function not covered" >me</span>rcator.bbox(x, y, zoom).join(',');
  tr<span class="cstat-no" title="statement not covered" >y {</span>
    if (debug) {
      return new DOMParser().parseFromString(await fetchStub(), 'text/xml');
    }
    let response = await fetch(
      `https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`,
    );
    response = handleErrors(response);
    const te<span class="missing-if-branch" title="if path not taken" >I</span>xt = await response.text();
<span class="cstat-no" title="statement not covered" ></span>
    const parser = new DOMParser();
    const xml = parser.parseFromString(text, 'text/xml');
    return xml;
  } catch (e) {
    console.error(e);
    throw e;
  }
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Oct 13 2017 13:43:03 GMT+0530 (IST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
