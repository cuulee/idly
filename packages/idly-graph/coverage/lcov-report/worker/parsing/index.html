<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for worker/parsing</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> worker/parsing
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">59.59% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>87/146</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">32.79% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>20/61</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">56.67% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>17/30</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">59.4% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>79/133</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="fetch.ts"><a href="fetch.ts.html">fetch.ts</a></td>
	<td data-value="77.78" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 77%;"></div><div class="cover-empty" style="width:23%;"></div></div></td>
	<td data-value="77.78" class="pct medium">77.78%</td>
	<td data-value="36" class="abs medium">28/36</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="11" class="abs high">10/11</td>
	<td data-value="77.78" class="pct medium">77.78%</td>
	<td data-value="9" class="abs medium">7/9</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="30" class="abs high">25/30</td>
	</tr>

<tr>
	<td class="file high" data-value="fixtures.ts"><a href="fixtures.ts.html">fixtures.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="nodeToFeature.ts"><a href="nodeToFeature.ts.html">nodeToFeature.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

<tr>
	<td class="file low" data-value="parser.ts"><a href="parser.ts.html">parser.ts</a></td>
	<td data-value="29.17" class="pic low"><div class="chart"><div class="cover-fill" style="width: 29%;"></div><div class="cover-empty" style="width:71%;"></div></div></td>
	<td data-value="29.17" class="pct low">29.17%</td>
	<td data-value="72" class="abs low">21/72</td>
	<td data-value="4.76" class="pct low">4.76%</td>
	<td data-value="42" class="abs low">2/42</td>
	<td data-value="31.25" class="pct low">31.25%</td>
	<td data-value="16" class="abs low">5/16</td>
	<td data-value="27.94" class="pct low">27.94%</td>
	<td data-value="68" class="abs low">19/68</td>
	</tr>

<tr>
	<td class="file high" data-value="wayToFeature.ts"><a href="wayToFeature.ts.html">wayToFeature.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="24" class="abs high">24/24</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="23" class="abs high">23/23</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Oct 13 2017 13:43:03 GMT+0530 (IST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
