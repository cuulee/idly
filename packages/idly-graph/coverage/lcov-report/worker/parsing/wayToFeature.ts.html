<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for worker/parsing/wayToFeature.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">worker/parsing</a> wayToFeature.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>24/24</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>8/8</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>4/4</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>23/23</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58</td><td class="line-coverage quiet"><span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">11x</span>
<span class="cline-any cline-yes">11x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">25x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">17x</span>
<span class="cline-any cline-yes">11x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { EntityTable, Node, OsmGeometry, Way } from 'idly-common/lib/osm/structures';
import * as turfLineString from 'turf-linestring';
import * as turfPolygon from 'turf-polygon';
&nbsp;
/**
 * @REVISIT this func also needs to handle modifiedGraph
 *  for now i removed it so i can test caching and anyway
 *  not yet implemented modified ways
 */
&nbsp;
export function wayCombiner(
  way: Way,
  table: EntityTable,
  existingProps: { 'osm_basic--geometry': OsmGeometry; [key: string]: string },
) {
  // @TOFIX code duplication
  // @REVISIT this osm_basic injection, sems hacks
  const existingGeometry = existingProps['osm_basic--geometry'];
  if (!existingGeometry) {
    throw new Error('geometry not found in existing props');
  }
  const nodes = getCoordsFromTable(table, way.nodes);
  return {
    id: way.id,
    ...wayToLineString(existingGeometry, nodes),
    properties: {
      ...existingProps,
      id: way.id,
    },
  };
}
&nbsp;
export function getCoordsFromTable(
  table: EntityTable,
  nodes: Way['nodes'],
): number[][] {
  return nodes.map(n =&gt; {
    if (!table.has(n)) throw new Error('node not found ' + n);
    const node = table.get(n) as Node;
    return [node.loc.lon, node.loc.lat];
  });
}
&nbsp;
/**
 * graphNode comes from graph.node
 *  planning to use it for memoization.
 *  every ms counts. hehehe.
 */
export function wayToLineString(geom: OsmGeometry, nodeCoords: number[][]) {
  if (geom === OsmGeometry.LINE) {
    return turfLineString(nodeCoords, {});
  } else if (geom === OsmGeometry.AREA) {
    return turfPolygon([nodeCoords], {});
  } else {
    throw new Error('not a matching geometry provided for way');
  }
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Oct 13 2017 13:43:03 GMT+0530 (IST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
