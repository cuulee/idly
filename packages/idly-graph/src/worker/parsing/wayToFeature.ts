import { EntityTable, Node, OsmGeometry, Way } from 'idly-common/lib/osm/structures';
import * as turfLineString from 'turf-linestring';
import * as turfPolygon from 'turf-polygon';

/**
 * @REVISIT this func also needs to handle modifiedGraph
 *  for now i removed it so i can test caching and anyway
 *  not yet implemented modified ways
 */

export function wayCombiner(
  way: Way,
  table: EntityTable,
  existingProps: { 'osm_basic--geometry': OsmGeometry; [key: string]: string },
) {
  // @TOFIX code duplication
  // @REVISIT this osm_basic injection, sems hacks
  const existingGeometry = existingProps['osm_basic--geometry'];
  if (!existingGeometry) {
    throw new Error('geometry not found in existing props');
  }
  const nodes = getCoordsFromTable(table, way.nodes);
  return {
    id: way.id,
    ...wayToLineString(existingGeometry, nodes),
    properties: {
      ...existingProps,
      id: way.id,
    },
  };
}

export function getCoordsFromTable(
  table: EntityTable,
  nodes: Way['nodes'],
): number[][] {
  return nodes.map(n => {
    if (!table.has(n)) throw new Error('node not found ' + n);
    const node = table.get(n) as Node;
    return [node.loc.lon, node.loc.lat];
  });
}

/**
 * graphNode comes from graph.node
 *  planning to use it for memoization.
 *  every ms counts. hehehe.
 */
export function wayToLineString(geom: OsmGeometry, nodeCoords: number[][]) {
  if (geom === OsmGeometry.LINE) {
    return turfLineString(nodeCoords, {});
  } else if (geom === OsmGeometry.AREA) {
    return turfPolygon([nodeCoords], {});
  } else {
    throw new Error('not a matching geometry provided for way');
  }
}
