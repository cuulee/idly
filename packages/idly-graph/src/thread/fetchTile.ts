import { entityTableGen } from 'idly-common/lib/osm/entityTableGen';
import { Entity, EntityTable, ParentWays } from 'idly-common/lib/osm/structures';

import { calculateParentWays } from '../misc/calculateParentWays';
import { fetchTileXml } from './fetchTileXml';
import { xmlToEntities } from './xmlToEntities';

export async function fetchTile(
  x: number,
  y: number,
  z: number,
): Promise<{
  readonly entities: Entity[];
  readonly entityTable: EntityTable;
  readonly parentWays: ParentWays;
}> {
  const xml = await fetchTileXml(x, y, z);
  const entities = xmlToEntities(xml);
  const entityTable = entityTableGen(entities);
  const parentWays = calculateParentWays(entityTable);
  return { entities, entityTable, parentWays };
}
