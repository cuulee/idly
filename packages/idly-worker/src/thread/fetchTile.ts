import { Tile } from 'idly-common/lib/geo/tile';
import { entityTableGen } from 'idly-common/lib/osm/entityTableGen';
import { smartParser } from 'idly-faster-osm-parser';

import { calculateParentWays } from '../misc/calculateParentWays';
import { tileId } from '../misc/tileId';
import { TileData, TilesDataTable } from '../operations/operationsTypes';
import { fetchTileXml } from './fetchTileXml';

export async function fetchTile(
  x: number,
  y: number,
  z: number
): Promise<TileData> {
  const xml = await fetchTileXml(x, y, z);
  const entities = smartParser(xml);

  const entityTable = entityTableGen(entities);
  const parentWays = calculateParentWays(entityTable);
  return { entities, entityTable, parentWays };
}

// TOFIX handle not caching when project.rejects
export function cacheFetchTile(
  tilesDataTable: TilesDataTable,
  tiles: Tile[]
): {
  readonly tilesDataTable: TilesDataTable;
  readonly tilesData: Array<Promise<TileData>>;
} {
  const tilesData = tiles.map(({ x, y, z }) => {
    const str = tileId({ x, y, z });
    const cached = tilesDataTable.get(str);
    if (cached) {
      return cached;
    }
    const res = fetchTile(x, y, z);
    tilesDataTable = tilesDataTable.set(str, res);
    return res;
  });
  return { tilesDataTable, tilesData };
}
