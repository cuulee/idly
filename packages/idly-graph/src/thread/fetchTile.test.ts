import { miniXML3 } from '../worker/parsing/fixtures';
import { fetchTile } from './fetchTile';

declare var global: any;
// tslint:disable no-expression-statement no-object-mutation

describe('fetchTile', () => {
  test('gives an xml', async () => {
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve({
          id: '123',
          async text(): Promise<any> {
            return miniXML3;
          },
          ok: true,
        });
      });
    });
    const xml = await fetchTile(17, 17, 18);
    expect(xml).toMatchSnapshot();
  });
  test('should fail if status > 400', async () => {
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve({
          id: '123',
          ok: false,
          statusText: 'errored',
          async text(): Promise<any> {
            return miniXML3;
          },
        });
      });
    });
    expect.assertions(1);
    fetchTile(17, 17, 18).catch(e => expect(e.message).toEqual('errored'));
  });
});
