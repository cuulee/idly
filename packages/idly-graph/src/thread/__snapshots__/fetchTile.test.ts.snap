// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchTile basic test 1`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n1",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": "[]",
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n2",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": "[]",
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n3",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": "[]",
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w1",
    "nodes": Array [
      "n1",
      "n2",
    ],
    "tags": "[]",
    "type": "way",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w2",
    "nodes": Array [
      "n1",
      "n3",
    ],
    "tags": "[]",
    "type": "way",
  },
]
`;
